lib/fluent/plugin/filter_analyze_config.rb

```diff
@@ -0,0 +1,275 @@
+# Copyright 2020 Google Inc. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+require 'fileutils'
+require 'fluent/config'
+require 'fluent/config/v1_parser'
+require 'set'
+
+require_relative 'monitoring'
+
+module Fluent
+  # Fluentd filter plugin to analyze configuration usage.
+  #
+  # For documentation on inspecting parsed configuration elements, see
+  # https://www.rubydoc.info/github/fluent/fluentd/Fluent/Config/Element
+  class AnalyzeConfigFilter < Filter
+    include Fluent::Config
+    Fluent::Plugin.register_filter('analyze_config', self)
+
+    module Constants
+      # Built-in plugins that are ok to reference in metrics.
+      KNOWN_PLUGINS = {
+        'filter' => Set[
+          'geoip',
+          'grep',
+          'parser',
+          'record_transformer',
+          'stdout',
+        ],
+        'match' => Set[
+          'copy',
+          'elasticsearch',
+          'exec',
+          'exec_filter',
+          'file',
+          'forward',
+          'http',
+          'kafka',
+          'mongo',
+          'mongo_replset',
+          'null',
+          'relabel',
+          'rewrite_tag_filter',
+          'roundrobin',
+          's3',
+          'secondary_file',
+          'stdout',
+          'webhdfs',
+        ],
+        'source' => Set[
+          'dummy',
+          'exec',
+          'forward',
+          'http',
+          'monitor_agent',
+          'syslog',
+          'tail',
+          'tcp',
+          'udp',
+          'unix',
+          'windows_eventlog',
+        ]
+      }.freeze
+
+      # These are the config params for which we want usage metrics.
+      GOOGLE_CONFIG_PARAMS = %w(
```

Comment:
We have some other plugins that we own as well. Maybe `s/GOOGLE_CONFIG_PARAMS/OUTPUT_GOOGLE_CLOUD_CONFIG_PARAMS`?

Oh this reminds me that: can we add a check for whether `detect_exceptions` is being used as well (https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions)? Ideally also the params it support, but that's less urgent than to know whether the plugin is being used at all.

Sample config:
```
<match **>
  @type detect_exceptions
  remove_tag_prefix foo
  message log
  languages java, python
  multiline_flush_interval 0.1
</match>
```

Reply:
Sounds good. I added param metrics for that plugin as well. Any others plugins?

I renamed the variable to GOOGLE_PLUGIN_PARAMS, let me know if you prefer another name.